go/vt/vtorc/discovery/queue.go

```diff
@@ -33,85 +33,78 @@ import (
 	"vitess.io/vitess/go/vt/vtorc/config"
 )
 
-// Queue contains information for managing discovery requests
-type Queue struct {
-	sync.Mutex
+// queueItem represents an item in the discovery.Queue.
+type queueItem struct {
+	PushedAt time.Time
+	Key      string
+}
 
-	name         string
-	done         chan struct{}
-	queue        chan string
-	queuedKeys   map[string]time.Time
-	consumedKeys map[string]time.Time
+// Queue is an ordered queue with deduplication.
+type Queue struct {
+	mu       sync.Mutex
+	enqueued map[string]struct{}
+	queue    chan queueItem
 }
 
-// CreateQueue allows for creation of a new discovery queue
-func CreateQueue(name string) *Queue {
+// NewQueue creates a new queue.
+func NewQueue() *Queue {
 	return &Queue{
-		name:         name,
-		queuedKeys:   make(map[string]time.Time),
-		consumedKeys: make(map[string]time.Time),
-		queue:        make(chan string, config.DiscoveryQueueCapacity),
+		enqueued: make(map[string]struct{}),
+		queue:    make(chan queueItem, config.DiscoveryQueueCapacity),
+	}
+}
+
+// setKeyCheckEnqueued returns true if a key is already enqueued, if
+// not the key will be marked as enqueued and false is returned.
+func (q *Queue) setKeyCheckEnqueued(key string) (alreadyEnqueued bool) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	_, alreadyEnqueued = q.enqueued[key]
+	if !alreadyEnqueued {
+		q.enqueued[key] = struct{}{}
 	}
+	return alreadyEnqueued
 }
 
-// QueueLen returns the length of the queue (channel size + queued size)
+// QueueLen returns the length of the queue.
 func (q *Queue) QueueLen() int {
-	q.Lock()
-	defer q.Unlock()
+	q.mu.Lock()
+	defer q.mu.Unlock()
 
-	return len(q.queue) + len(q.queuedKeys)
+	return len(q.queue) + len(q.enqueued)
```

Comment:
This is actually incorrect. The length of of enqueued should be enough.

Because `enqueued` will have all the elements that are queued and also the ones that are currently running. By adding the length of queue, we are countint the former elements twice.

Reply:
Yeah, this is what the previous code did but it's pretty confusing

It seems to be measuring how much is sitting in the channel buffer plus what is queued. The size of the channel buffer is interesting, but confusing to include

I might add a new metric for the channel buffer in a new PR and a method like `.BufferLen() int`. What do you think?

Reply:
Updated the code